daemon/state.go

```diff
@@ -221,118 +221,134 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
+				scopedLog.Warn("Endpoint to restore has been deleted")
 				return
 			}
-			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
-			// Filter the restored labels with the new daemon's filter
-			l, _ := labels.FilterLabels(ep.OpLabels.IdentityLabels())
-			ep.RUnlock()
-
-			identity, _, err := cache.AllocateIdentity(context.Background(), l)
-			if err != nil {
-				scopedLog.WithError(err).Warn("Unable to restore endpoint")
-				epRegenerated <- false
-			}
 
 			// Wait for initial identities and ipcache from the
 			// kvstore before doing any policy calculation for
 			// endpoints that don't have a fixed identity or are
 			// not well known.
+			identity := ep.SecurityIdentity
 			if !identity.IsFixed() && !identity.IsWellKnown() {
 				cache.WaitForInitialIdentities(context.Background())
 				ipcache.WaitForInitialSync()
 			}
 
-			if err := ep.LockAlive(); err != nil {
-				scopedLog.Warn("Endpoint to restore has been deleted")
-				return
-			}
-
-			ep.LogStatusOKLocked(endpoint.Other, "Synchronizing endpoint labels with KVStore")
-
-			if ep.SecurityIdentity != nil {
-				if oldSecID := ep.SecurityIdentity.ID; identity.ID != oldSecID {
-					log.WithFields(logrus.Fields{
-						logfields.EndpointID:              ep.ID,
-						logfields.IdentityLabels + ".old": oldSecID,
-						logfields.IdentityLabels + ".new": identity.ID,
-					}).Info("Security identity for endpoint is different from the security identity restored for the endpoint")
-
-					// The identity of the endpoint being
-					// restored has changed. This can be
-					// caused by two main reasons:
-					//
-					// 1) Cilium has been upgraded,
-					// downgraded or the configuration has
-					// changed and the new version or
-					// configuration causes different
-					// labels to be considered security
-					// relevant for this endpoint.
-					//
-					// Immediately using the identity may
-					// cause connectivity problems if this
-					// is the first endpoint in the cluster
-					// to use the new identity. All other
-					// nodes will not have had a chance to
-					// adjust the security policies for
-					// their endpoints. Hence, apply a
-					// grace period to allow for the
-					// update. It is not required to check
-					// any local endpoints for potential
-					// outdated security rules, the
-					// notification of the new security
-					// identity will have been received and
-					// will trigger the necessary
-					// recalculation of all local
-					// endpoints.
-					//
-					// 2) The identity is outdated as the
-					// state in the kvstore has changed.
-					// This reason would justify an
-					// immediate use of the new identity
-					// but given the current identity is
-					// already in place, it is also correct
-					// to continue using it for the
-					// duration of a grace period.
-					time.Sleep(defaults.IdentityChangeGracePeriod)
-				}
+			if _, wait := deferredRegenerations[ep]; wait {
```

Comment:
Yep, and that's a big part of what causes #7358 (along with the uninitialized `ep.SecurityIdentity.LabelArray`), because the security identity isn't properly set up and it generates an l4policy using the wrong labels.